Framework/PythonInterface/mantid/api/src/Exports/IFunction.cpp

```diff
@@ -35,6 +36,32 @@ PyObject *getCategories(IFunction &self) {
   return registered;
 }
 
+/// An object for constructing a shared_ptr that won't ever delete its pointee
+class NoDeleting {
+public:
+  /// Does nothing
+  void operator()(void *) {}
+  /// Does nothing
+  void operator()(const void *) {}
+};
+
+//------------------------------------------------------------------------------------------------------
+/**
+* Something that makes IFunction return to python a composite function
+* for Product function, Convolution or
+* any similar superclass of composite function.
+* @param self :: Enables it to be called as a member function on the
+* IFunction class
+*/
+Mantid::API::CompositeFunction_sptr castToCompositeFunction(IFunction &self) {
```

Comment:
@KarlPalmen I think the issue is to enforce the associative property of the product operator:
```
( f * g ) * h == f * (g * h) == f * g * h
```
To me the most appropriate venue is to enforce this in the `ProductFunction` C++ class. It  requires overriding `addFunction()` and `replaceFunction()` of parent class `CompositeFunction`. For instance, if we are adding a `ProductFunction`, then we loop `addFunction()` over its components functions. 
Once this is done, writing any of the three previous expressions will result in the same object.